### ISSUE TITLE
`caml_string_is_c_safe` and `caml_unix_check_path` aren't used correctly in OCaml 5

### ISSUE THREAD

**Issue by dra27** [ocaml/ocaml#11326]
In discussion with @OlivierNicole about #11325, I think we have an issue with the way we use `caml_string_is_c_safe` and therefore `caml_sys_check_path` and `caml_unix_check_path`.

In OCaml 4.x, the runtime lock is held which provides an atomicity guarantee between the call to `caml_string_is_c_safe` and duplicating it or passing it to a syscall. The only way the string could be mutated would involve a violation of the runtime lock, which is "not a concern".

In OCaml 5.x, although it requires the use of an unsafe feature to mutate the string, it would be possible for another domain to mutate the string without violating the GC rules, which is much more of a concern.

Most uses of these functions are followed by a copy, with these exceptions:
- `caml_unix_inet_addr_of_string`
- `caml_unix_getgrnam`
- `caml_unix_gethostbyname`
- `caml_unix_getprotobyname`
- `caml_unix_getpwnam`
- `caml_unix_getservbyname`
- `caml_unix_getservbyport`
- `caml_unix_initgroups`
- `caml_unix_realpath`
- `caml_unix_get_sockaddr`
- `caml_unix_spawn` (both implementations)

I propose:
- Those functions should be switched to copy their arguments, not for GC-safety, but for parallel safety
- In all cases, the duplication should be done first (it's not unsafe) and _then_ the C string should validated

**Comment by gasche** [comment]
Don´t we have the same problem at the OCaml level? If I call `Unix.inet_addr_of_string str`, with your proposal there is still a race between `inet_addr_of_string` copying the string and another thread mutating the string. Your proposal guarantees that the C runtime calls will not see files with \0 characters in them, but not that the string provided by OCaml is the one intended by the caller.

Going back to the source of these checks, #6945, I see that the rationale for those checks is not really the correctness of the C syscalls themselves, but rather a potential mismatch between the C interpretation of the string and higher-level checks performed by the programming language: the [Null Byte Injection](http://projects.webappsec.org/w/page/13246949/Null%20Byte%20Injection) documentation gives this Perl example:

```perl
$buffer = $ENV{'QUERY_STRING'};
$buffer =~ s/%([a-fA-F0-9][a-fA-F0-9])/pack("C", hex($1))/eg;
$fn = '/home/userx/data/' .$buffer. '.jpg';
open (FILE,"<$fn");
```

> Exploitation:
> Normal Mode: http://www.example.host/read.pl?page=userphoto.jpg
> Attacking Mode: http://www.example.host/read.pl?page=../../../../etc/passwd%00jpg

As far as I can tell, in the scenario that you have in mind (adversarial mutations to a string shared between domains), the protection that you propose does not protect from the attack suggested in this example: instead of inserting a null byte (which is correctly prevented by your proposal), the malicious domain can simply rewrite the string from `userphoto.jpg` to `/etc/passwd` concurrently, after the check has passed.

**Comment by dra27** [comment]
Indeed - my proposal only addresses the C level

**Comment by dra27** [comment]
I wasn't thinking of this from an adversarial point of view, but more from correctness - the uses of `caml_string_is_c_safe` assume that having checked the string it is then "safe" from that point onwards, which no longer holds. The security-conscious user should still be worrying about the mutation of strings like this at the OCaml level (and therefore copying them before performing other validation), but the bug in the C code should - for me, at least - be fixed, regardless.

**Comment by dbuenzli** [comment]
I don't understand. Aren't strings immutable nowadays ?

**Comment by nojb** [comment]
To play devil's advocate: the present issue is only an issue if using unsafe features of the language. But in that case, all bets are off, right?

**Comment by gasche** [comment]
My impression:
- the checks are designed to fix security issues coming from not-paranoid-enough high-level checks on strings that don't include "suspicious null bytes" in their model
- in a sequential world, the checks correctly guarantee the absence of this issue
- in a concurrent world, the checks don't guarantee the absence of this issue, but we could keep them as a form of "defense in depth" (security theater?) or fix them
- in a concurrent world, many more attacks are possibly on shared mutable strings, to the point that null-byte injection are not a concern anymore in the context of high-level checks

Three options seem reasonable:
1. The checks don't provide extra security guarantees anymore, let's remove them.
2. We care about the checks *or* it's simpler to do the clearly-right thing, let's fix the checks.
3. The checks cannot provide security guarantees but they still make sense in some scenarios (including all purely-sequential programs), let's keep things as-is.

If the performance cost of additional string copies was clearly neglectible, I would vote for (2). However I have the impression that some functions may be more performance-sensitive than that (I could see some services calling `inet_addr_of_string` on a frequent basis for example?). Maybe (1) is also a sensible option, as it simplifies the codebase instead of keeping cruft around that we are not sure serves a purpose.

**Comment by gasche** [comment]
> The security-conscious user should still be worrying about the mutation of strings like this at the OCaml level (and therefore copying them before performing other validation), but the bug in the C code should - for me, at least - be fixed, regardless.

That's a good point: if the user is aware of concurrent-unsafe-mutation attacks but not of null-byte-injection attacks, they could carefully copy strings locally but then forget to check for null bytes, and in this scenario the null-byte check is useful. But in this case the fix that your propose is not useful, right?

**Comment by dra27** [comment]
> > The security-conscious user should still be worrying about the mutation of strings like this at the OCaml level (and therefore copying them before performing other validation), but the bug in the C code should - for me, at least - be fixed, regardless.
> 
> That's a good point: if the user is aware of concurrent-unsafe-mutation attacks but not of null-byte-injection attacks, they could carefully copy strings locally but then forget to check for null bytes, and in this scenario the null-byte check is useful. But in this case the fix that your propose is not useful, right?

That's right - in this case, the string is already copied, so the only parallel mutation would be one the caller had explicitly set-up themselves.

**Comment by dra27** [comment]
> To play devil's advocate: the present issue is only an issue if using unsafe features of the language. But in that case, all bets are off, right?

In general, I agree - the possible difference here (hence opening issue before PR!) is that it was not possible to circumvent this check with unsafe features in OCaml 4.x but it has now become possible to do so in OCaml 5.x.

**Comment by dbuenzli** [comment]
> * the checks are designed to fix security issues coming from not-paranoid-enough high-level checks on strings that don't include "suspicious null bytes" in their model

They are here to test for a mismatch between the notion of string in C – which are null terminated and thus cannot include null bytes – and OCaml's notion of string, which are also null terminated and can be used directly as C string with the caveat that they can also include null bytes in their content.

> * in a concurrent world, the checks don't guarantee the absence of this issue, but we could keep them as a form of "defense in depth" (security theater?) or fix them

Could you please explain to people like me, who don't have their head deep into this new concurrent world, how exactly these tests are now broken assuming you respect OCaml stance on `string`s that is that they are immutable ?

Before imagining "attacks", it's better to first define a threat model. In my threat model the program itself is not part of the attacker and so there's nothing to defend against here. 

The only thing you may want to defend against are programmers that try to be smart, but I'm afraid there won't be a foolproof answer to that.

**Comment by dbuenzli** [comment]
> Maybe (1) is also a sensible option, as it simplifies the codebase instead of keeping cruft around that we are not sure serves a purpose.

Btw. since this doesn't seem to be clearly understood.

These checks are here because without them you violate the programmer's simple and reasonable expectation that *all* of the bytes you pass from OCaml as a string are going to to be seen by the corresponding C function.

Removing these checks would be insane. 

How many of your programs actually check your cli arguments for null bytes before giving them to these functions ? How many URL fragments, JSON strings are simply decoded and passed further after some steps to these functions without checking them for null bytes ?

**Comment by gasche** [comment]
> Could you please explain to people like me, who don't have their head deep into this new concurrent world, how exactly these tests are now broken assuming you respect OCaml stance on strings that is that they are immutable ?

I think the idea is to assume that string mutations can still happen in practice, for example through `Bytes.unsafe_to_string`. It is reasonable for many purposes to assume that programs misusing this feature are buggy and therefore unworthy of consideration, but it would also be reasonable I think to have a mental model where the language (currently) "encourages you to treat strings as immutable", but does not actually enforce this property.

**Comment by gasche** [comment]
@dbuenzli if I understand, you are thinking of a reasonably correct program that is passed invalid inputs. What should the specification of `Unix.inet_addr_of_string` say about input strings that contain a null byte? Failing, you argue, is better than silently truncating the input. That makes a lot of sense, thanks! In this scenario, "doing nothing" (my option (3)) is the right choice I think.

**Comment by dbuenzli** [comment]
> Failing, you argue, is better than silently truncating the input.

Yes, because the truncated data will never be what you intended to give to the function[^1]. As a result you may get denial of service, but not worse stuff which can happen when people start [to play with this](https://owasp.org/www-community/attacks/Embedding_Null_Code).

> (my option (3)) is the right choice I think.

Personally I think so too. And it's not only about security, it's also about making sound bindings to C functions that take C strings that you represent in OCaml as `string`s. You have an abstraction mismatch it seems quite vicious to silently hide it to the clients of the API.


[^1]: For the same reason you should never simply drop decoding errors when you do best-effort UTF text decoding. You always need to add something e.g. `Uchar.rep` to indicate there was something there, otherwise, depending on where decodes happen, attackers might be able to circumvent filters etc.

**Comment by jberdine** [comment]
FWIW: +1 to failing instead of truncating (that is, silently doing the wrong thing) for this and other instances of passing strings from OCaml to C.

**Comment by xavierleroy** [comment]
I'm in strong agreement with @dbuenzli.  Strings are immutable in OCaml, which gives nice guarantees both for sequential and for concurrent programs.  And we're not giving any guarantees for programs that use `Bytes.unsafe_of_string` to mutate strings concurrently with uses of these strings, just like we're not giving any guarantees for programs that use `Obj.magic`.  If programmers want to shoot themselves in the foot, that's their problem.

The existing checks for null characters in the middle of strings are as useful and relevant as ever to protect against malicious user-provided data.

In perfect hindsight, I agree with @dra27 that it would have been nice to combine the null character check with the "copy an OCaml string from the OCaml heap to malloc-ed storage" operation.  But there's nothing wrong with the code as it is today, and we have many more urgent things to do for OCaml 5.

For these reasons, I move to close this issue.

**Comment by jberdine** [comment]
Not to open Pandora's box, and I have not investigated the module dependency order, but would it make sense to move `Bytes.unsafe_of_string` and friends to `Obj` for OCaml 5?

**Comment by nojb** [comment]
> Not to open Pandora's box, and I have not investigated the module dependency order, but would it make sense to move `Bytes.unsafe_of_string` and friends to `Obj` for OCaml 5?

Personally in favour; but since these functions are part of the public API this change needs to go through the usual deprecation cycle. Probably a good idea to open an issue for it though.

**Comment by dbuenzli** [comment]
I have mixed feeling about that. Depends on which "friends" we are talking about here.

The need to be able to locally allocate a bytes value, crunch to define its content by mutation and then escape it from the function as an immutable string via `Bytes.unsafe_to_string` without making a new copy is real and users will do it. If that function ends up in `Obj` it gets harder to sell the "don't ever use `Obj`" line.

**Comment by dra27** [comment]
I'm happy to close - @gasche?

**Comment by gasche** [comment]
Sounds okay to me as well -- we collectively convinced ourselves that the existing check is important, but that the proposed change (with unknown performance impact) is not arguably required. In this case doing nothing is the easiest road to a reasonable solution.
